src/Storages/Hive/StorageHive.cpp

```diff
@@ -0,0 +1,666 @@
+        if (file->hasMinMaxIndex())
+        {
+            file->loadMinMaxIndex();
+            if (!hivefile_key_condition.checkInHyperrectangle(file->getMinMaxIndex()->hyperrectangle, hivefile_name_types.getTypes())
+                     .can_be_true)
+            {
+                LOG_DEBUG(log, "skip file:{} index:{}", file->getPath(), file->describeMinMaxIndex(file->getMinMaxIndex()));
+                return;
+            }
+        }
+
+        // Load sub-file level minmax index and apply
+        std::set<int> skip_splits;
+        if (file->hasSubMinMaxIndex())
+        {
+            file->loadSubMinMaxIndex();
+            const auto & sub_minmax_idxes = file->getSubMinMaxIndexes();
+            for (size_t i = 0; i < sub_minmax_idxes.size(); ++i)
+            {
+                if (!hivefile_key_condition.checkInHyperrectangle(sub_minmax_idxes[i]->hyperrectangle, hivefile_name_types.getTypes())
+                         .can_be_true)
+                {
+                    LOG_DEBUG(log, "skip split:{} in file {}", i, file->getPath());
+                    skip_splits.insert(i);
+                }
+            }
+            file->setSkipSplits(skip_splits);
+        }
+
+        {
+            std::lock_guard lock{hive_files_mutex};
+            hive_files.push_back(file);
+        }
+    };
+
+    ThreadPool pool{num_streams};
+    if (!partitions.empty())
+    {
+        const auto partition_names = partition_name_types.getNames();
+        const auto partition_types = partition_name_types.getTypes();
+
+        for (const auto & p : partitions)
+        {
+            auto f = [&]()
+            {
+                // Skip partition "__HIVE_DEFAULT_PARTITION__"
+                bool has_default_partition = false;
+                for (const auto & value : p.values)
+                {
+                    if (value == "__HIVE_DEFAULT_PARTITION__")
+                    {
+                        has_default_partition = true;
+                        break;
+                    }
+                }
+                if (has_default_partition)
+                {
+                    //LOG_DEBUG(log, "skip partition:__HIVE_DEFAULT_PARTITION__");
+                    return;
+                }
+
+                std::vector<Range> ranges;
+                WriteBufferFromOwnString wb;
+                if (p.values.size() != partition_names.size())
+                    throw Exception(
+                        fmt::format("Partition value size not match, expect {}, but got {}", partition_names.size(), p.values.size()),
+                        ErrorCodes::INVALID_PARTITION_VALUE);
+
+                for (size_t i = 0; i < p.values.size(); ++i)
+                {
+                    if (i != 0)
+                        writeString(",", wb);
+                    writeString(p.values[i], wb);
+                }
+                writeString("\n", wb);
```

Comment:
Writing local files in `read` methods looks like a bad idea, as well as creating a threadpool. This method can be called more than once for single read btw depending on query processing stage.
What this `f` lambda is supposed to do? Why have to write local files?

Reply:
Please notice that we are writing to memory, not to local files.  

`f` lambda collect files to read under single partiton, in `f` we apply partition minmax index if the query has partition key condition. BTW, I renamed `f` to `process_partition`